### ISSUE TITLE
Enhancement: prefer-enum-initializers values must be the same as keys

### ISSUE THREAD

**Issue by benediktms** [typescript-eslint/typescript-eslint#7149]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/prefer-enum-initializers

### Description

enums in Typescript can sometimes be unintuitive to use, and don't always produce the results that are expected, mainly, I belive, due to the fact that by default they are objects that assign number values to their keys. 

Here a detailed description on why enums can be unintuitive: https://www.youtube.com/watch?v=jjMbPt_H3RQ&t=358s

I propose a config option to the `prefer-enum-initializers` rule that enforces that enum values must always match their keys.

### Fail

```typescript
enum Foo {
  Bar,
  Baz
}

enum Foo {
  Bar = 0,
  Baz = 1
}

enum Foo {
  Bar = 'hello',
  Baz = 'world'
}

enum Foo {
  Bar = 'BAR',
  Baz = 'BAZ'
}
```

### Pass

```typescript
enum Foo {
  Bar = 'Bar',
  Baz = 'Baz'
}
```

**Comment by bradzacher** [comment]
> Here a detailed description on why enums can be unintuitive

I strongly disagree with really every single argument set forth within the video and it conflates a lot of ideas which is what makes things confusing.

- "If you think of an enum like an object" => enums aren't an object though, so of course the models don't line up.
    - Yes, they are an object at runtime (because there's no better representation) but type-wise they are not objects.
- The video misrepresents enum values as just being a type alias for their concrete value => enums aren't an alias they're a (pseudo)nominal type on top of the underlying value.
    - It's better to think of `LogLevel.DEBUG` as having type `'DEBUG' & { __brand: unknown }` - not the type `'DEBUG'`.
    - This is why you can assign `LogLevel.DEBUG` to `'DEBUG'` - but not the other way around because TS allows you to widen the type from the enum to the literal (or even the base `string` type) - but not the other way around.
- They are scared off of `const` enums even though the handbook clearly states solid workaround to the pitfalls it lists.
    - Also TS itself **HEAVILY** leverages `const` enums internally
    - The referenced section even explicitly states "Unlike inlining enums from other projects, inlining a project’s own enums is not problematic" - it clearly says that they're safe to use!
- They recommend `as const` objects with a type-shadow alias as a simpler solution
    - Not sure how it's simpler if it requires more code for the declaration and advanced concepts like type/value shadowing.
    - You also additionally need `typeof` to resolve the type of the member, rather than directly referencing the member.
    - It's no longer a (pseudo)nominal type - which means you don't need use the value via the "enum".
        - This is bad for codebase maintenance - if you ever want to change an enum member's value you need to manually find-and-replace all of the string literals instead of just changing the enum member's value.

TL;DR
- enums are not objects - don't think of them like that at all and you'll go very far.
- enum members are not aliases - [they're completely unique types](https://www.typescriptlang.org/play?#code/KYOwrgtgBAYg9nKBvKBBKBeKByV2oC+AUKJFAEICGATsmpjnoUUQPStQDGc11wnAFyjBePKJQDOUAQAtgUCMAgAjEVJrywIAJYBHMPIEBPAA7AJAGigATcye0D5MygDdtIAObS5UCZUVQ1FoC2gEulAA2BkTcIBJCAB4AXBQ0AHToWPBwGQDcLOxcPHyCwqK0kt7yiioiUNrqUFp6BtKm8gDuMtqcMvWNEmDKxmZQcABmjPgAtOIg1lBGcGBclCAgcEIgNNRwHeJto26U4hIS2h4giiACMXBxQkYp2RkMuNj5QA). Think of them as a `unique symbol` and you'll go very far.

**Comment by bradzacher** [comment]
All that being said - I'm not quite sure I understand the motivation behind your request.
How does making the value and key the same make enums more clear?
How would enforcing this standard be useful to the broader community?

**Comment by bradzacher** [comment]
> but they can still get coerced into a primitive when performing comparisons

[This is why we strongly recommend one or both of our `strict-boolean-expressions` and `no-unnecessary-conditions` lint rules](https://typescript-eslint.io/play/#ts=5.0.3&jsx=true&sourceType=module&code=KYOwrgtgBAYg9nKBvAUFKBBANGqAhKAXigHI8SUBfFGgYzhAGcAXKAQyNgQDoMBuFAEsAZlAAUbAJTIo1FPSasARp3hxueASPFLpSWSiA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZfKsugIwHs+SAIZN6AD2LQUySnyaoMkRNGh9okcGAC+ILUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA) because they will catch and prevent these cases of implicit coercion in conditional locations.

We also have the rule [`no-mixed-enums`](https://typescript-eslint.io/rules/no-mixed-enums) which [similarly prevents bad code of implicit `0` enum values](https://typescript-eslint.io/play/#ts=5.0.3&jsx=true&sourceType=module&code=KYOwrgtgBAYg9nKBvAUFKBBANGqAhKAXigHI8SUBfFIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oFtKAPRABN6TWN1QZIiaNA7RI4MAF8QSoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA). Paired with `prefer-enum-initializers` it means that you can't have a mixed number/string enum AND you must have an explicit initialiser for every enum member (preventing implicit `0`s and only allowing explicit ones).

We also have the rule [`no-unsafe-enum-comparisons`](https://typescript-eslint.io/rules/no-unsafe-enum-comparison) which will [prevent explicit comparisons against non-enum values](https://typescript-eslint.io/play/#ts=5.0.3&jsx=true&sourceType=module&code=KYOwrgtgBAYg9nKBvAUFKBBANGqAhKAXigHI8SUBfFFAYzhAGcAXKAQyNgQDoMBuFAEsAZlAAUHQlKgAGAJTIo1OgxZQARp3hxueASPGapxMiQVIlKIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tieQEMAZonpNYAW1pkOY4nyrIOTdFETRoHaJHBgAviB1A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA).

There are lots of option that help make enums safer to use!

> I think enforcing enums values to have the same name as the key is another layer of predictability that you can add to your codebase. You always know what the value of the enum will be without having to double check.

I definitely understand that - but I disagree with an option that forces you to use string enums AND force you to use a string which exactly matches the key.

I've seen a lot of codebases everywhere which make good use of number enums for reduced memory usage, and I've seen a lot of codebases use naming conventions like either "enum members must be UPPER_SNAKE_CASE" or "enum members must be PascalCase" - which means you can't always match the enum name to the enum value.

For example your usecase specifies that this code is invalid:
```ts
enum Foo {
  Bar = 'BAR',
  Baz = 'BAZ'
}
```
However in many codebases I've seen this is exactly the correct style!

---

I think what you're really looking for is banning number enums altogether due to their chance of being coerced to false. [That's entirely doable right now!](https://typescript-eslint.io/linting/troubleshooting/#how-can-i-ban-specific-language-feature)
[Example playground](https://typescript-eslint.io/play#ts=5.0.3&jsx=true&sourceType=module&showAST=es&code=PTAEFMCdIe0gocA7ArgW1AMRjUBveUUAIQENIAaQk0gL3gF95FUMzJ9rtcBeUABipF2oPgEZ+koTVqjQAWjGNmIUKQA260AHMcAExboZnYeTkAifuellZfc2POMgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WnqdgBbWs0r5KAQ0YAvXqgyRe0APbRIAGnDZITZbU7J8VMvkQATWskIsJAD3RgA2ougq16zFuxRkiJCdUHSAAVAGUAUQFBAFlEQQAjXjAAPjA0QQl8MgALFAAKABkxXilHaAkAd2SAXgAGAF0AOlFxKUpZaA8AVSYJaEJw2w4UZEplJmamMUkZXgBKDS9tQRGJAHNEIIA5IUToMEQo5DA+xDAzSmQpeGUK80glgF96rUeQR6A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)

**Comment by Zamiell** [comment]
First off, we should mention that the `prefer-enum-initializers` rule is mostly having to do with disallowing the "auto-computed" feature of enums. The style of initializers is a separate concern, so it feels better to put this check in a completely separate rule. Perhaps `consistent-enum-values`.

I also would like to see the enhancement specified in the OP. I would even go one step further and say that the proposed style is objectively better. Here's a short discussion on that.

To start with, we should mention that in a pure TypeScript program, the values of an enum should be superfluous - they should not affect the runtime behavior of the program in any way. Thus, the format with which you name your enum values is purely a stylistic concern.

Since the values of enums don't matter, it follows that the most maintainable and less-prone-to-error solution would be to use the auto-computed feature of TypeScript, like this:

```ts
enum Foo {
  Bar,
  Baz,
}
```

Unfortunately, the auto-computed feature uses numeric values instead of string values, and as previously mentioned in this thread, the TypeScript compiler is not as strict with number enums as it is with string enums.

Thus, in order to get the maximum guarantees of safety in our programs, we want to use string enums instead of number enums, and therefore cannot use the auto-compute feature.

So then, what is the next simplest and easiest-to-maintain solution? I think it would be simply copying the values of the the enum exactly, like this:

```ts
enum Foo {
  Bar = "Bar",
  Baz = "Baz",
}
```

This is the stylistic solution recommended in the OP, and one that is easily enforced by a lint rule.

Above, Brad mentions that he has seen the following style in many existing TypeScript codebases:

```ts
enum Foo {
  Bar = "BAR",
  Baz = "BAZ",
}
```

Here, we are applying an arbitrary transformation to the enum members, turning them into uppercase values. But what exactly is the point of this extra step / extra transformation? Remember that the actual enum values are completely superfluous to the functioning of the program (at least, in a pure TypeScript program). So we aren't "getting" anything by enforcing this extra transformation.

I think that the reason that Brad has seen this style in other codebases is purely an artifact of legacy JavaScript code. In JavaScript, before TypeScript existed, it was a common convention to name constants in capital letters, like this:

```js
var MY_VALUE = 123;
```

By naming constants with capital letters and normal variables with camelCase, we could get more sane programs and prevent classes of bugs. Even though the language itself did not support immutability, programmers would just know that they were never supposed to be modifying variables that were in capital letters. Similar to constant variables, enum values are also supposed to be immutable, so this convention was carried over into TypeScript land.

Of course, in TypeScript compiler, it is never possible to change the value of a TypeScript enum in the first place, so this convention doesn't serve any purpose anymore. It is simply a legacy of older times.

**Comment by Zamiell** [comment]
Above, Brad mentions that some code bases specifically use number enums for the purposes of reducing memory usage (in return for having less safe code). How much memory would you really be saving here? I imagine that you would have to have quite a lot of enums in your TypeScript program for this to ever classify as a non micro-optimization. I suspect that in the vast majority of TypeScript programs, the marginal memory cost of converting to number enums would not ever be measurable.

**Comment by bradzacher** [comment]
> How much memory would you really be saving here?

Well if you do the math - in JS a `number` is stored with 64-bits (8 bytes).
So unless your member name is 8 characters or less (which generally they end up being longer) - you're using `n-8` extra bytes per storage instance of the enum.

TypeScript has the [`SyntaxKind` enum](https://github.com/microsoft/TypeScript/blob/6c942fa2c98fa157c0335a1f9820c28f93dbc022/src/compiler/types.ts#L41-L495) which it uses like the `AST_NODE_TYPES` enum in our project. It has some 450 members averaging a member name length of 15.2 characters (max = 44, min = 5).

So on average we're wasting 7 bytes per usage of `SyntaxKind`.

If you typecheck a TypeScript project (eg just run `tsc`) then this will create hundreds of millions of AST nodes - which is hundreds of millions of instances of `SyntaxKind`.

Just doing that quick math using the average (which isn't truly correct because each `SyntaxKind` isn't used equally - but good enough to get the gist) - if you use strings then:
- 1m * 15.2b = 15mb.
- 10m * 15.2b = 145mb.
- 100m * 15.2b = 1.45gb.

That's a lot of ram just to store some object identifiers!

But if you use numbers:
- 1m * 8b = 8mb.
- 10m * 8b = 75mb.
- 100m * 8b = 745mb.

So we've saved this much ram by using numbers:
- 1m = 7mb
- 10m = 70mb
- 100m = 705mb

This isn't just theoretical either - this is a real world usecase where numbers can save so much ram at scale! It is for sure not the common usecase by any means - most usecases will likely only have dozens or maybe hundreds of instances of an enum - so for most usecases it really doesn't matter!

But there's also the case of over-the-wire cost to consider.
Imagine instead you're running an application that serves 100 million users - big but not *massive*.
Now you're in exactly the same boat! If you use string enums to communicate with your API then you could be wasting gigabytes of bandwidth every single day - that equates to real $$$!

**Comment by bradzacher** [comment]
> Here, we are applying an arbitrary transformation to the enum members, turning them into uppercase values. But what exactly is the point of this extra step / extra transformation?

Sometimes it's required to communicate with some external API (either within JS or over-the-wire via HTTP). Sometimes it's just a legacy thing where names have changed but values haven't.

But it's ultimately a stylistic choice within the codebase - the code isn't any better or worse for using a style where the enum name doesn't match the enum value. In your opinion it might be - but in someone else's opinion it might not be. That's the thing about code style - it's opinionated!

To be clear - I'm not saying you're wrong! I personally prefer having the style of things matching! But it's just not the only right answer

**Comment by Zamiell** [comment]
@benediktms I made some new rules:

- [consistent-enum-values](https://github.com/IsaacScript/isaacscript/blob/main/packages/eslint-plugin-isaacscript/docs/rules/consistent-enum-values.md)
- [no-number-enums](https://github.com/IsaacScript/isaacscript/blob/main/packages/eslint-plugin-isaacscript/docs/rules/no-number-enums.md)

You should use these two rules in addition to [`@typescript-eslint/prefer-enum-initializers`](https://typescript-eslint.io/rules/prefer-enum-initializers/) and [`@typescript-eslint/no-unsafe-enum-comparison](https://typescript-eslint.io/rules/no-unsafe-enum-comparison) for maximum safety.

If Brad wants, I can submit `consistent-enum-values` to this repo as a PR. As for `no-number-enums`, it sounds like the consensus right now on the `@typescript-eslint` team is that there are too many use-cases in the wild for number enums, so making an official rule to ban them is too heavy handed.

**Comment by benediktms** [comment]
@Zamiell yes the argument you posted sums up my thoughts exactly. As for the convention of using string vs numbers as enum values. I belive (I have no evidence) that most projects that worry about enum correctness probably are not at the scale where they need to worry about variable space complexity, but as @bradzacher said, the benefit (in context of the JS ecosystem, having very fragmented style opinions) here is that rules can be turned on or off.

@Zamiell I've tested your plugins, and they work perfectly for my use case, thank you so much for sharing 😃

**Comment by bradzacher** [comment]
Given the new alternatives I think we can consider this closed.

If the community really gets behind the above rules we can consider revisiting this to upstream them and broaden their reach.
